File: AssembleAlgorithms.py
from LinearSystem import LinearSystem
from FieldsHolder import FieldArray, DimType,MAX_DIM, FieldNames
from MeshObject import MeshObject
from enum import Enum

class Boundary(Enum):
    LEFT = 0
    RIGHT = 1
    TOP = 2
    BOTTOM = 3

class AssembleSystemBase:
    def __init__(self, name, fieldsHolder, linearSystem, meshObject):
        self.name_ = name
        self.fieldsHolder_ : FieldArray = fieldsHolder
        self.linearSystem_ : LinearSystem = linearSystem
        self.myMeshObject_ : MeshObject = meshObject
    def zero(self):
        self.linearSystem_.zero()
    def assemble(self):
        raise NotImplementedError("Must implement assemble() in subclass")
    def __repr__(self):
        return f"AssembleSystemBase(name='{self.name_}', field='{self.fieldsHolder_.get_name()}', linearSystem='{self.linearSystem_}')"

class AssembleCellVectorTimeTerm(AssembleSystemBase):
    def __init__(self, name, fieldsHolderNew, fieldsHolderOld, linearSystem, meshObject, dt):
        super().__init__(name, fieldsHolderNew, linearSystem, meshObject)
        self.dt_ = dt
        self.fieldHolderOld_ = fieldsHolderOld
        assert fieldsHolderNew.get_type() == DimType.VECTOR, "AssembleCellVectorTimeTerm requires a vector field for state NEW"
        assert fieldsHolderOld.get_type() == DimType.VECTOR, "AssembleCellVectorTimeTerm requires a vector field for state OLD"
    def assemble(self):
        for cell in self.myMeshObject_.get_cells():
            cellID = cell.get_flat_id()
            cellVolume = cell.get_volume()
            for comp in range(self.fieldsHolder_.get_num_components()):
                lhsFactor = cellVolume/self.dt_
                rowIndex = cellID*self.fieldsHolder_.get_num_components() + comp
                value = (self.fieldsHolder_.get_data()[rowIndex] - self.fieldHolderOld_.get_data()[rowIndex])
                self.linearSystem_.add_lhs(rowIndex, rowIndex, -lhsFactor)
                self.linearSystem_.add_rhs(rowIndex, value*lhsFactor)

class AssembleCellVectorPressureGradientToLinSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolderNew, linearSystem, meshObject):
        super().__init__(name, fieldsHolderNew, linearSystem, meshObject)
        assert fieldsHolderNew.get_type() == DimType.VECTOR, "AssembleCellVectorPressureGradient requires a vector field for state NEW"
        assert fieldsHolderNew.get_name() == FieldNames.GRAD_PRESSURE.value, "AssembleCellVectorPressureGradient requires a field named 'grad_pressure' for the pressure gradient values"
    def assemble(self):
        for cell in self.myMeshObject_.get_cells():
            cellID = cell.get_flat_id()
            cellVolume = cell.get_volume()
            for comp in range(self.fieldsHolder_.get_num_components()):
                rowIndex = cellID*self.fieldsHolder_.get_num_components() + comp
                value = (self.fieldsHolder_.get_data()[rowIndex])*cellVolume
                self.linearSystem_.add_rhs(rowIndex, value)

class AssembleInteriorVectorAdvectionToLinSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolder, linearSystem, meshObject, diffusionCoeff=1.0):
        super().__init__(name, fieldsHolder, linearSystem, meshObject)
        self.diffusionCoeff_ = diffusionCoeff
        assert fieldsHolder.get_type() == DimType.VECTOR, "AssembleInteriorVectorAdvectionToLinSystem requires a vector field"
    def assemble(self):
        # Loop over internal faces and assemble contributions to the linear system
        for face in self.myMeshObject_.get_internal_faces():
            leftCellID = face.get_left_cell()
            rightCellID = face.get_right_cell()
            faceFlux = face.massFlux_
            mdot_L = (faceFlux + abs(faceFlux))/2.0
            mdot_R = (faceFlux - abs(faceFlux))/2.0
            fieldLeft = [self.fieldsHolder_.get_data()[leftCellID*MAX_DIM + i] for i in range(MAX_DIM)]
            fieldRight = [self.fieldsHolder_.get_data()[rightCellID*MAX_DIM + i] for i in range(MAX_DIM)]
            rhs = [fieldLeft[i]*mdot_L + fieldRight[i]*mdot_R for i in range(MAX_DIM)]

            # Add contribution to the RHS/LHS
            for i in range(MAX_DIM):
                rowIndexLeft = leftCellID*MAX_DIM+i
                rowIndexRight = rightCellID*MAX_DIM+i
                self.linearSystem_.add_rhs(rowIndexLeft, rhs[i])
                self.linearSystem_.add_rhs(rowIndexRight, -rhs[i])
                self.linearSystem_.add_lhs(rowIndexLeft, rowIndexLeft, -mdot_L)  
                self.linearSystem_.add_lhs(rowIndexLeft, rowIndexRight, -mdot_R)  
                self.linearSystem_.add_lhs(rowIndexRight, rowIndexLeft, mdot_L)  
                self.linearSystem_.add_lhs(rowIndexRight, rowIndexLeft, mdot_R)  

    def __repr__(self):
        return super().__repr__()


class AssembleInteriorVectorDiffusionToLinSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolder, linearSystem, meshObject, diffusionCoeff=1.0):
        super().__init__(name, fieldsHolder, linearSystem, meshObject)
        self.diffusionCoeff_ = diffusionCoeff
        assert fieldsHolder.get_type() == DimType.VECTOR, "AssembleInteriorVectorDiffusionToLinSystem requires a vector field"
    def assemble(self):
        # Loop over internal faces and assemble contributions to the linear system
        for face in self.myMeshObject_.get_internal_faces():
            leftCellID = face.get_left_cell()
            rightCellID = face.get_right_cell()
            faceArea = face.get_area()
            normal = face.get_normal_vector()
            cellLeft = self.myMeshObject_.get_cell_by_flat_id(leftCellID)
            cellRight = self.myMeshObject_.get_cell_by_flat_id(rightCellID)
            distance = [cellRight.get_centroid()[i] - cellLeft.get_centroid()[i] for i in range(MAX_DIM)]  # Vector from left cell to right cell
            normalLength = normal[0]*distance[0] + normal[1]*distance[1]  # Dot product of normal and distance vector
            lhsFactor = faceArea/normalLength*self.diffusionCoeff_
            gradDotArea = [(self.fieldsHolder_.get_data()[rightCellID*MAX_DIM + i] - self.fieldsHolder_.get_data()[leftCellID*MAX_DIM + i])*lhsFactor for i in range(MAX_DIM)]

            # Add contribution to the RHS/LHS
            for i in range(MAX_DIM):
                self.linearSystem_.add_rhs(leftCellID*MAX_DIM + i, -gradDotArea[i])
                self.linearSystem_.add_rhs(rightCellID*MAX_DIM + i, gradDotArea[i])
                self.linearSystem_.add_lhs(leftCellID*MAX_DIM + i, rightCellID*MAX_DIM + i, lhsFactor)  
                self.linearSystem_.add_lhs(leftCellID*MAX_DIM + i, leftCellID*MAX_DIM + i, -lhsFactor)  
                self.linearSystem_.add_lhs(rightCellID*MAX_DIM + i, rightCellID*MAX_DIM + i, -lhsFactor)  
                self.linearSystem_.add_lhs(rightCellID*MAX_DIM + i, leftCellID*MAX_DIM + i, lhsFactor)  

    def __repr__(self):
        return super().__repr__()
    
class AssembleDirichletBoundaryVectorDiffusionToLinSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolder, linearSystem, meshObject, boundaryType, boundaryValue, diffusionCoeff=1.0):
        super().__init__(name, fieldsHolder, linearSystem, meshObject)
        assert type(boundaryType) == Boundary, "boundaryType must be an instance of the Boundary enum"
        assert fieldsHolder.get_type() == DimType.VECTOR, "AssembleDirichletBoundaryVectorDiffusionToLinSystem requires a vector field"
        assert type(boundaryValue) == list and len(boundaryValue) == MAX_DIM, "boundaryValue must be a list of length equal to the number of dimensions"
        self.myFaceIterable_ = None
        if boundaryType == Boundary.LEFT:
            self.myFaceIterable_ = meshObject.get_left_boundary()
        elif boundaryType == Boundary.RIGHT:
            self.myFaceIterable_ = meshObject.get_right_boundary()
        elif boundaryType == Boundary.TOP:
            self.myFaceIterable_ = meshObject.get_top_boundary()
        elif boundaryType == Boundary.BOTTOM:
            self.myFaceIterable_ = meshObject.get_bottom_boundary()  
        assert self.myFaceIterable_ is not None, "Invalid boundary type specified"

        self.diffusionCoeff_ = diffusionCoeff
        self.boundaryValue_ = boundaryValue  # The Dirichlet value (e.g., Temperature at wall)

    def assemble(self):
        """Loop over boundary faces and assemble diffusive flux."""
        for face in self.myFaceIterable_:
            cellID = face.get_left_cell()
            faceArea = face.get_area()
            normal = face.get_normal_vector()
            
            owningCell = self.myMeshObject_.get_cell_by_flat_id(cellID)
            
            # Vector from cell centroid to face centroid
            distanceVec = [face.get_face_center()[i] - owningCell.get_centroid()[i] for i in range(2)]
            
            # This is 'd_f' in the flux equation: J = -D * (phi_boundary - phi_cell) / d_f
            normalDistance = (normal[0]*distanceVec[0] + normal[1]*distanceVec[1])
            lhsFactor = (self.diffusionCoeff_ * faceArea) / normalDistance
            gradDotArea = [(self.boundaryValue_[i] - self.fieldsHolder_.get_data()[cellID*MAX_DIM + i])*lhsFactor for i in range(MAX_DIM)]

            for i in range(MAX_DIM):
                self.linearSystem_.add_rhs(cellID*MAX_DIM+i, -gradDotArea[i])
                self.linearSystem_.add_lhs(cellID*MAX_DIM+i, cellID*MAX_DIM+i, -lhsFactor)

class AssembleInteriorPressurePoissonSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolder, linearSystem, meshObject, dt):
        super().__init__(name, fieldsHolder, linearSystem, meshObject)
        self.dt_ = dt
        assert fieldsHolder.get_type() == DimType.SCALAR, "AssembleInteriorPressurePoissonSystem requires a scalar field"
        assert fieldsHolder.get_name() == FieldNames.PRESSURE.value, "AssembleInteriorPressurePoissonSystem requires a field named 'Pressure' for the pressure values"
    def assemble(self):
        # Loop over internal faces and assemble contributions to the linear system
        for face in self.myMeshObject_.get_internal_faces():
            leftCellID = face.get_left_cell()
            rightCellID = face.get_right_cell()
            faceArea = face.get_area()
            normal = face.get_normal_vector()
            cellLeft = self.myMeshObject_.get_cell_by_flat_id(leftCellID)
            cellRight = self.myMeshObject_.get_cell_by_flat_id(rightCellID)
            distance = [cellRight.get_centroid()[i] - cellLeft.get_centroid()[i] for i in range(MAX_DIM)]  # Vector from left cell to right cell
            normalLength = normal[0]*distance[0] + normal[1]*distance[1]  # Dot product of normal and distance vector
            lhsFactor = self.dt_*faceArea/normalLength
            massFlux = face.massFlux_

            # Add contribution to the RHS
            self.linearSystem_.add_rhs(leftCellID, massFlux) 
            self.linearSystem_.add_rhs(rightCellID, -massFlux)
            # Add contribution to the LHS
            self.linearSystem_.add_lhs(leftCellID, rightCellID, lhsFactor)  
            self.linearSystem_.add_lhs(leftCellID, leftCellID, -lhsFactor)  
            self.linearSystem_.add_lhs(rightCellID, rightCellID, -lhsFactor)  
            self.linearSystem_.add_lhs(rightCellID, leftCellID, lhsFactor)  

class AssembleInteriorScalarDiffusionToLinSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolder, linearSystem, meshObject, diffusionCoeff=1.0):
        super().__init__(name, fieldsHolder, linearSystem, meshObject)
        self.diffusionCoeff_ = diffusionCoeff
        assert fieldsHolder.get_type() == DimType.SCALAR, "AssembleInteriorScalarDiffusionToLinSystem requires a scalar field"
    def assemble(self):
        # Loop over internal faces and assemble contributions to the linear system
        for face in self.myMeshObject_.get_internal_faces():
            leftCellID = face.get_left_cell()
            rightCellID = face.get_right_cell()
            faceArea = face.get_area()
            normal = face.get_normal_vector()
            cellLeft = self.myMeshObject_.get_cell_by_flat_id(leftCellID)
            cellRight = self.myMeshObject_.get_cell_by_flat_id(rightCellID)
            distance = [cellRight.get_centroid()[i] - cellLeft.get_centroid()[i] for i in range(MAX_DIM)]  # Vector from left cell to right cell
            normalLength = normal[0]*distance[0] + normal[1]*distance[1]  # Dot product of normal and distance vector
            lhsFactor = faceArea/normalLength*self.diffusionCoeff_
            gradDotArea = (self.fieldsHolder_.get_data()[rightCellID] - self.fieldsHolder_.get_data()[leftCellID])*lhsFactor

            # Add contribution to the RHS
            self.linearSystem_.add_rhs(leftCellID, -gradDotArea) 
            self.linearSystem_.add_rhs(rightCellID, gradDotArea)
            # Add contribution to the LHS
            self.linearSystem_.add_lhs(leftCellID, rightCellID, lhsFactor)  
            self.linearSystem_.add_lhs(leftCellID, leftCellID, -lhsFactor)  
            self.linearSystem_.add_lhs(rightCellID, rightCellID, -lhsFactor)  
            self.linearSystem_.add_lhs(rightCellID, leftCellID, lhsFactor)  

    def __repr__(self):
        return super().__repr__()
    
class AssembleDirichletBoundaryScalarDiffusionToLinSystem(AssembleSystemBase):
    def __init__(self, name, fieldsHolder, linearSystem, meshObject, boundaryType, boundaryValue, diffusionCoeff=1.0):
        super().__init__(name, fieldsHolder, linearSystem, meshObject)
        assert type(boundaryType) == Boundary, "boundaryType must be an instance of the Boundary enum"
        assert fieldsHolder.get_type() == DimType.SCALAR, "AssembleDirichletBoundaryScalarDiffusionToLinSystem requires a scalar field"
        self.myFaceIterable_ = None
        if boundaryType == Boundary.LEFT:
            self.myFaceIterable_ = meshObject.get_left_boundary()
        elif boundaryType == Boundary.RIGHT:
            self.myFaceIterable_ = meshObject.get_right_boundary()
        elif boundaryType == Boundary.TOP:
            self.myFaceIterable_ = meshObject.get_top_boundary()
        elif boundaryType == Boundary.BOTTOM:
            self.myFaceIterable_ = meshObject.get_bottom_boundary()  
        assert self.myFaceIterable_ is not None, "Invalid boundary type specified"

        self.diffusionCoeff_ = diffusionCoeff
        self.boundaryValue_ = boundaryValue  # The Dirichlet value (e.g., Temperature at wall)

    def assemble(self):
        """Loop over boundary faces and assemble diffusive flux."""
        for face in self.myFaceIterable_:
            cellID = face.get_left_cell()
            faceArea = face.get_area()
            normal = face.get_normal_vector()
            
            owningCell = self.myMeshObject_.get_cell_by_flat_id(cellID)
            
            # Vector from cell centroid to face centroid
            distanceVec = [face.get_face_center()[i] - owningCell.get_centroid()[i] for i in range(2)]
            
            # This is 'd_f' in the flux equation: J = -D * (phi_boundary - phi_cell) / d_f
            normalDistance = (normal[0]*distanceVec[0] + normal[1]*distanceVec[1])
            lhsFactor = (self.diffusionCoeff_ * faceArea) / normalDistance
            gradDotArea = (self.boundaryValue_ - self.fieldsHolder_.get_data()[cellID])*lhsFactor
            self.linearSystem_.add_rhs(cellID, -gradDotArea)
            self.linearSystem_.add_lhs(cellID, cellID, -lhsFactor)

File: CPP_Impl/src/AssembleAlgorithms.cpp
#include "AssembleAlgorithms.h"
#include <stdexcept>
#include <cmath>
#include <iostream>
#include <vector>

namespace Vinci4D {

AssembleCellVectorTimeTerm::AssembleCellVectorTimeTerm(
    const std::string& name, FieldArray* fieldsHolderNew, FieldArray* fieldsHolderOld,
    LinearSystem* linearSystem, MeshObject* meshObject, double dt)
    : AssembleSystemBase(name, fieldsHolderNew, linearSystem, meshObject),
      dt_(dt), fieldHolderOld_(fieldsHolderOld) {
    
    if (fieldsHolderNew->getType() != DimType::VECTOR) {
        throw std::runtime_error("AssembleCellVectorTimeTerm requires a vector field for state NEW");
    }
    if (fieldsHolderOld->getType() != DimType::VECTOR) {
        throw std::runtime_error("AssembleCellVectorTimeTerm requires a vector field for state OLD");
    }
}

void AssembleCellVectorTimeTerm::assemble() {
    const auto& cells = meshObject_->getCells();
    auto& dataNew = fieldsHolder_->getData();
    const auto& dataOld = fieldHolderOld_->getData();
    int numComponents = fieldsHolder_->getNumComponents();
    
    for (const auto& cell : cells) {
        int cellID = cell.getFlatId();
        double cellVolume = cell.getVolume();
        
        for (int comp = 0; comp < numComponents; ++comp) {
            double lhsFactor = cellVolume / dt_;
            int rowIndex = cellID * numComponents + comp;
            double value = dataNew[rowIndex] - dataOld[rowIndex];
            
            linearSystem_->addLhs(rowIndex, rowIndex, -lhsFactor);
            linearSystem_->addRhs(rowIndex, value * lhsFactor);
        }
    }
}

AssembleInteriorVectorDiffusionToLinSystem::AssembleInteriorVectorDiffusionToLinSystem(
    const std::string& name, FieldArray* fieldsHolder, LinearSystem* linearSystem,
    MeshObject* meshObject, double diffusionCoeff)
    : AssembleSystemBase(name, fieldsHolder, linearSystem, meshObject),
      diffusionCoeff_(diffusionCoeff) {
    
    if (fieldsHolder->getType() != DimType::VECTOR) {
        throw std::runtime_error("AssembleInteriorVectorDiffusionToLinSystem requires a vector field");
    }
}

void AssembleInteriorVectorDiffusionToLinSystem::assemble() {
    const auto& internalFaces = meshObject_->getInternalFaces();
    auto& data = fieldsHolder_->getData();
    
    int numCells = meshObject_->getNumCells();
    
    for (const auto* face : internalFaces) {
        int leftCellID = face->getLeftCell();
        int rightCellID = face->getRightCell();
        
        // Debug: check if cell IDs are valid
        if (leftCellID < 0 || leftCellID >= numCells || 
            rightCellID < 0 || rightCellID >= numCells) {
            std::cerr << "Invalid cell IDs in internal face: leftID=" << leftCellID 
                      << " rightID=" << rightCellID << " (numCells=" << numCells << ")" << std::endl;
            throw std::runtime_error("Invalid cell ID in internal face");
        }
        
        // Get cell centroids
        const Cell* leftCell = meshObject_->getCellByFlatId(leftCellID);
        const Cell* rightCell = meshObject_->getCellByFlatId(rightCellID);
        
        if (!leftCell || !rightCell) {
            throw std::runtime_error("Invalid cell ID in internal face");
        }
        
        auto leftCentroid = leftCell->getCentroid();
        auto rightCentroid = rightCell->getCentroid();
        auto normal = face->getNormalVector();
        
        // Compute distance between cell centers along normal direction
        double distance = 0.0;
        for (int i = 0; i < MAX_DIM; ++i) {
            distance += (rightCentroid[i] - leftCentroid[i]) * normal[i];
        }
        
        double faceArea = face->getArea();
        double diffusionFactor = diffusionCoeff_ * faceArea / distance;
        
        // Assemble for each component
        for (int comp = 0; comp < MAX_DIM; ++comp) {
            int rowLeft = leftCellID * MAX_DIM + comp;
            int rowRight = rightCellID * MAX_DIM + comp;
            
            double phiLeft = data[rowLeft];
            double phiRight = data[rowRight];
            double flux = diffusionFactor * (phiRight - phiLeft);
            
            // Contribution to left cell
            linearSystem_->addLhs(rowLeft, rowLeft, -diffusionFactor);
            linearSystem_->addLhs(rowLeft, rowRight, diffusionFactor);
            linearSystem_->addRhs(rowLeft, -flux);
            
            // Contribution to right cell
            linearSystem_->addLhs(rowRight, rowRight, -diffusionFactor);
            linearSystem_->addLhs(rowRight, rowLeft, diffusionFactor);
            linearSystem_->addRhs(rowRight, flux);
        }
    }
}

AssembleCellVectorPressureGradientToLinSystem::AssembleCellVectorPressureGradientToLinSystem(
    const std::string& name, FieldArray* fieldsHolder, LinearSystem* linearSystem,
    MeshObject* meshObject)
    : AssembleSystemBase(name, fieldsHolder, linearSystem, meshObject) {
    
    if (fieldsHolder->getType() != DimType::VECTOR) {
        throw std::runtime_error("AssembleCellVectorPressureGradient requires a vector field");
    }
}

void AssembleCellVectorPressureGradientToLinSystem::assemble() {
    const auto& cells = meshObject_->getCells();
    auto& data = fieldsHolder_->getData();
    int numComponents = fieldsHolder_->getNumComponents();
    
    for (const auto& cell : cells) {
        int cellID = cell.getFlatId();
        double cellVolume = cell.getVolume();
        
        for (int comp = 0; comp < numComponents; ++comp) {
            int rowIndex = cellID * numComponents + comp;
            double value = data[rowIndex] * cellVolume;
            linearSystem_->addRhs(rowIndex, value);
        }
    }
}

AssembleDirichletBoundaryVectorDiffusionToLinSystem::AssembleDirichletBoundaryVectorDiffusionToLinSystem(
    const std::string& name, FieldArray* fieldsHolder, LinearSystem* linearSystem,
    MeshObject* meshObject, Boundary boundaryType,
    const std::array<double, 2>& boundaryValue, double diffusionCoeff)
    : AssembleSystemBase(name, fieldsHolder, linearSystem, meshObject),
      boundaryType_(boundaryType), boundaryValue_(boundaryValue), diffusionCoeff_(diffusionCoeff) {
    
    if (fieldsHolder->getType() != DimType::VECTOR) {
        throw std::runtime_error("Boundary vector diffusion requires a vector field");
    }
    
    // Populate boundary faces based on boundary type
    if (boundaryType == Boundary::LEFT) {
        boundaryFaces_ = meshObject->getLeftBoundaryFaces();
    } else if (boundaryType == Boundary::RIGHT) {
        boundaryFaces_ = meshObject->getRightBoundaryFaces();
    } else if (boundaryType == Boundary::TOP) {
        boundaryFaces_ = meshObject->getTopBoundaryFaces();
    } else if (boundaryType == Boundary::BOTTOM) {
        boundaryFaces_ = meshObject->getBottomBoundaryFaces();
    }
}

void AssembleDirichletBoundaryVectorDiffusionToLinSystem::assemble() {
    auto& data = fieldsHolder_->getData();
    
    for (auto* face : boundaryFaces_) {
        int cellID = face->getLeftCell();
        if (cellID < 0 || cellID >= meshObject_->getNumCells()) continue;
        
        const Cell* cell = meshObject_->getCellByFlatId(cellID);
        if (!cell) continue;
        
        double faceArea = face->getArea();
        auto normal = face->getNormalVector();
        auto cellCentroid = cell->getCentroid();
        auto faceCentroid = face->getFaceCenter();
        
        // Distance from cell centroid to face centroid
        double distX = faceCentroid[0] - cellCentroid[0];
        double distY = faceCentroid[1] - cellCentroid[1];
        double normalDistance = normal[0] * distX + normal[1] * distY;
        
        double lhsFactor = (diffusionCoeff_ * faceArea) / normalDistance;
        
        for (int comp = 0; comp < MAX_DIM; ++comp) {
            int rowIdx = cellID * MAX_DIM + comp;
            double gradDotArea = (boundaryValue_[comp] - data[rowIdx]) * lhsFactor;
            linearSystem_->addRhs(rowIdx, -gradDotArea);
            linearSystem_->addLhs(rowIdx, rowIdx, -lhsFactor);
        }
    }
}

AssembleInteriorVectorAdvectionToLinSystem::AssembleInteriorVectorAdvectionToLinSystem(
    const std::string& name, FieldArray* fieldsHolder, LinearSystem* linearSystem,
    MeshObject* meshObject)
    : AssembleSystemBase(name, fieldsHolder, linearSystem, meshObject) {
    
    if (fieldsHolder->getType() != DimType::VECTOR) {
        throw std::runtime_error("Interior vector advection requires a vector field");
    }
}

void AssembleInteriorVectorAdvectionToLinSystem::assemble() {
    const auto& internalFaces = meshObject_->getInternalFaces();
    auto& data = fieldsHolder_->getData();
    
    for (const auto* face : internalFaces) {
        int leftCellID = face->getLeftCell();
        int rightCellID = face->getRightCell();
        
        if (leftCellID < 0 || rightCellID < 0) continue;
        if (leftCellID >= meshObject_->getNumCells() || rightCellID >= meshObject_->getNumCells()) continue;
        
        double massFlux = face->getMassFlux();
        double mdot_L = (massFlux + std::abs(massFlux)) / 2.0;  // Upwind for left
        double mdot_R = (massFlux - std::abs(massFlux)) / 2.0;  // Upwind for right
        
        for (int i = 0; i < MAX_DIM; ++i) {
            double fieldLeft = data[leftCellID * MAX_DIM + i];
            double fieldRight = data[rightCellID * MAX_DIM + i];
            double rhs = fieldLeft * mdot_L + fieldRight * mdot_R;
            
            int rowIndexLeft = leftCellID * MAX_DIM + i;
            int rowIndexRight = rightCellID * MAX_DIM + i;
            
            linearSystem_->addRhs(rowIndexLeft, rhs);
            linearSystem_->addRhs(rowIndexRight, -rhs);
            linearSystem_->addLhs(rowIndexLeft, rowIndexLeft, -mdot_L);
            linearSystem_->addLhs(rowIndexLeft, rowIndexRight, -mdot_R);
            linearSystem_->addLhs(rowIndexRight, rowIndexLeft, mdot_L);
            linearSystem_->addLhs(rowIndexRight, rowIndexRight, mdot_R);
        }
    }
}

AssembleInteriorPressurePoissonSystem::AssembleInteriorPressurePoissonSystem(
    const std::string& name, FieldArray* fieldsHolder, LinearSystem* linearSystem,
    MeshObject* meshObject, double dt)
    : AssembleSystemBase(name, fieldsHolder, linearSystem, meshObject), dt_(dt) {
    
    if (fieldsHolder->getType() != DimType::SCALAR) {
        throw std::runtime_error("Pressure Poisson requires a scalar field");
    }
}

void AssembleInteriorPressurePoissonSystem::assemble() {
    const auto& internalFaces = meshObject_->getInternalFaces();
    
    for (const auto* face : internalFaces) {
        int leftCellID = face->getLeftCell();
        int rightCellID = face->getRightCell();
        
        if (leftCellID < 0 || rightCellID < 0) continue;
        if (leftCellID >= meshObject_->getNumCells() || rightCellID >= meshObject_->getNumCells()) continue;
        
        const Cell* leftCell = meshObject_->getCellByFlatId(leftCellID);
        const Cell* rightCell = meshObject_->getCellByFlatId(rightCellID);
        if (!leftCell || !rightCell) continue;
        
        double faceArea = face->getArea();
        auto normal = face->getNormalVector();
        auto leftCent = leftCell->getCentroid();
        auto rightCent = rightCell->getCentroid();
        
        double distX = rightCent[0] - leftCent[0];
        double distY = rightCent[1] - leftCent[1];
        double normalLength = normal[0] * distX + normal[1] * distY;
        
        double lhsFactor = dt_ * faceArea / normalLength;
        double massFlux = face->getMassFlux();
        
        linearSystem_->addRhs(leftCellID, massFlux);
        linearSystem_->addRhs(rightCellID, -massFlux);
        linearSystem_->addLhs(leftCellID, rightCellID, lhsFactor);
        linearSystem_->addLhs(leftCellID, leftCellID, -lhsFactor);
        linearSystem_->addLhs(rightCellID, rightCellID, -lhsFactor);
        linearSystem_->addLhs(rightCellID, leftCellID, lhsFactor);
    }
}

void AssembleInteriorPressurePoissonSystem::assembleRhs() {
    const auto& internalFaces = meshObject_->getInternalFaces();
    
    for (const auto* face : internalFaces) {
        int leftCellID = face->getLeftCell();
        int rightCellID = face->getRightCell();
        double massFlux = face->getMassFlux();
        
        linearSystem_->addRhs(leftCellID, massFlux);
        linearSystem_->addRhs(rightCellID, -massFlux);
    }
}

} // namespace Vinci4D
